In [1]:
# Import

import pandas as pd
import numpy as np
import pymc as pm
import arviz as az
import json
import re
import os
import networkx as nx

In [2]:
def load_jsonc(filepath):
    """Load JSONC file (JSON with comments)"""
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()
    
    # Remove single-line comments (// ...)
    content = re.sub(r'//.*?$', '', content, flags=re.MULTILINE)
    
    # Remove multi-line comments (/* ... */)
    content = re.sub(r'/\*.*?\*/', '', content, flags=re.DOTALL)
    
    # Remove trailing commas before closing brackets/braces
    content = re.sub(r',\s*([}\]])', r'\1', content)
    
    return json.loads(content)

def load_json(filepath):
    """Load regular JSON file"""
    with open(filepath, 'r', encoding='utf-8') as f:
        return json.load(f)

def load_any(filepath):
    """Load either JSON or JSONC file based on extension"""
    if filepath.endswith('.jsonc'):
        return load_jsonc(filepath)
    else:
        return load_json(filepath)

In [3]:
# TASK 1: relocate age_weights to parameters.jsonc file
input_dir = './input_data'
stage_dirs = ['amd_sim_data', 'amd_sim_data_Early', 'amd_sim_data_Intermediate', 'amd_sim_data_Late-dry', 'amd_sim_data_Late-wet']

for stage_dir in stage_dirs:
    parameters = load_jsonc(f'{input_dir}/{stage_dir}/parameters.jsonc')
    print(f'{stage_dir:<30} parameters keys: {parameters.keys()}')

amd_sim_data                   parameters keys: dict_keys(['rr', 'f', 'i', 'age_weights', 'ages', 'p', 'r', 'pf', 'X'])
amd_sim_data_Early             parameters keys: dict_keys(['rr', 'f', 'i', 'age_weights', 'ages', 'p', 'r', 'pf', 'X'])
amd_sim_data_Intermediate      parameters keys: dict_keys(['rr', 'f', 'i', 'age_weights', 'ages', 'p', 'r', 'pf', 'X'])
amd_sim_data_Late-dry          parameters keys: dict_keys(['rr', 'f', 'i', 'age_weights', 'ages', 'p', 'r', 'pf', 'X'])
amd_sim_data_Late-wet          parameters keys: dict_keys(['rr', 'f', 'i', 'age_weights', 'ages', 'p', 'r', 'pf', 'X'])


In [4]:
# TASK 2: update Load based on region_id_graph
stage_dirs = ['amd_sim_data', 'amd_sim_data_Early', 'amd_sim_data_Intermediate', 'amd_sim_data_Late-dry', 'amd_sim_data_Late-wet']
stage = stage_dirs[2] # Intermediate

#------------------------------------------------------------------------------------------------------------#
# Read input data

input_data = pd.read_csv(f'{input_dir}/{stage}/input_data.csv')
output_template = pd.read_csv(f'{input_dir}/{stage}/output_template.csv')
parameters = load_any(f'{input_dir}/{stage}/parameters.jsonc')    # dict. {'p': {}, 'age_weights': [], 'ages: []}
hierarchy = load_any(f'{input_dir}/{stage}/hierarchy.json')       # dict of lists. {'nodes': [], 'edges': []}
nodes_to_fit = load_any(f'{input_dir}/{stage}/nodes_to_fit.json') # LIST of strings

print(f'number of rows: {len(input_data)}')
print(f'number of unique location_id: {input_data["location_id"].nunique()}')
print('--------------------------------')

#------------------------------------------------------------------------------------------------------------#
# Create region_graph
nodes = hierarchy['nodes']
name_to_id = {} # warning: this does not handle duplicate names
id_to_name = {}

region_id_graph = nx.DiGraph()

for node in nodes:
  name_to_id[node[0]] = node[1]['location_id']
  id_to_name[node[1]['location_id']] = node[0]

  # create region_id_graph
  region_id_graph.add_node(node[1]['location_id'],
                            level = node[1]['level'],
                            parent_id = node[1]['parent_id'],
                            name = node[0]
                          )

  my_id = node[1]['location_id']
  parent_id = node[1]['parent_id']
  if my_id != parent_id: # if my_id is not the root node
    region_id_graph.add_edge(parent_id, my_id)

print(f"number of nodes: {region_id_graph.number_of_nodes()}") 
print(f"number of edges: {region_id_graph.number_of_edges()}")

number of rows: 207
number of unique location_id: 18
--------------------------------
number of nodes: 233
number of edges: 232


In [5]:
print(id_to_name[422]) # IDs are not incremental

print(region_id_graph.nodes[422]['level'])
print(region_id_graph.nodes[422]['parent_id'])
print(region_id_graph.nodes[422]['name'])

United States Virgin Islands
3
104
United States Virgin Islands


In [6]:
# TASK 3: define "coords"

country_list = []
region_list = []
super_region_list = []

for node in hierarchy['nodes']:
  if node[1]['level'] == 3:
    country_list.append(node[1]['location_id'])
  elif node[1]['level'] == 2:
    region_list.append(node[1]['location_id'])
  elif node[1]['level'] == 1:
    super_region_list.append(node[1]['location_id'])
    
coords = {
    "country":      country_list,
    "region":       region_list,
    "super_region": super_region_list,
}

print(len(country_list))
print(len(region_list))
print(len(super_region_list))

204
21
7


In [7]:
# Utilizes id_to_name to print the name of the location
def describe():
        G = region_id_graph
        df = input_data
        for n in nx.dfs_postorder_nodes(G, 1):
            cnt = df['location_id'].eq(n).sum() + sum(G.nodes[c].get('cnt', 0) for c in G.successors(n))
            G.nodes[n]['cnt'] = int(cnt)
            G.nodes[n]['depth'] = nx.shortest_path_length(G, 1, n)
            
        for n in nx.dfs_preorder_nodes(G, 1):
            if G.nodes[n]['cnt'] > 0:
                print('  '*G.nodes[n]['depth'] + id_to_name[n], G.nodes[n]['cnt'])

# describe()

def keep():
    pass
    # Suggestion: filter input_data during "LOAD"

def filter_input_data_by_data_type(input_data: pd.DataFrame, data_type: str) -> pd.DataFrame:
        if not input_data.empty:
            return input_data[input_data['data_type'] == data_type]
        return input_data

# since our input_data only has 'p' data, this will return the same input_data
filter_input_data_by_data_type(input_data, 'p').head() 

area  location_id         stage  stage_id     sex  sex_id  year_id   
0  Netherlands           89  Intermediate         5    Male       1     1990  \
1  Netherlands           89  Intermediate         5  Female       2     1990   
2  Netherlands           89  Intermediate         5    Male       1     1990   
3  Netherlands           89  Intermediate         5  Female       2     1990   
4  Netherlands           89  Intermediate         5    Male       1     1990   

   age_start  age_end  effective_sample_size     value  standard_error   
0         55       64                 1418.0  0.040903        0.005260  \
1         55       64                 1802.0  0.033851        0.004260   
2         65       74                 1382.0  0.072359        0.006969   
3         65       74                 1865.0  0.036997        0.004371   
4         75       84                  796.0  0.103015        0.010774   

      x_sdi     x_tob data_type  upper_ci  lower_ci  age_weights  
0  0.794612  0.434156         p       NaN       NaN          NaN  
1  0.794612  0.383810         p       NaN       NaN          NaN  
2  0.794612  0.434156         p       NaN       NaN          NaN  
3  0.794612  0.383810         p       NaN       NaN          NaN  
4  0.794612  0.434156         p       NaN       NaN          NaN

In [8]:
# process.asr() | parameters
# ------------------------------------------------------------
data_type                = 'p'
reference_area           = 'Global'
reference_sex            = 'Both'
reference_year           = 'all'
mu_age                   = None
mu_age_parent            = None
sigma_age_parent         = None
rate_type                = 'neg_binom'
lower_bound              = None
interpolation_method     = 'linear'
include_covariates       = True
zero_re                  = False
# ------------------------------------------------------------

In [9]:
# process.asr() | local variables
# ------------------------------------------------------------
ages = np.array(parameters['ages'], dtype=np.float64)
age_weights = np.array(parameters['age_weights'], dtype=np.float64)
data = filter_input_data_by_data_type(input_data, data_type)
lb_data =filter_input_data_by_data_type(input_data, lower_bound) if lower_bound else None
params_of_data_type = parameters.get(data_type, {})

# check: mu_age_parent & sigma_age_parent
#  if either mu_age_parent or sigma_age_parent is NaN, set them to None
if (isinstance(mu_age_parent, np.ndarray) and np.isnan(mu_age_parent).any()) or \
    (isinstance(sigma_age_parent, np.ndarray) and np.isnan(sigma_age_parent).any()):

    mu_age_parent = None
    sigma_age_parent = None
# ------------------------------------------------------------
print(f'ages: {ages} | type: {type(ages)} | ages.dtype: {ages.dtype}')

ages: [ 2.  3.  4.  5.  6.  7.  8.  9. 10. 11. 12. 13. 14. 15. 16. 17. 18. 19.
 20. 21. 22. 23. 24. 25. 26. 27. 28. 29. 30. 31. 32. 33. 34. 35. 36. 37.
 38. 39. 40. 41. 42. 43. 44. 45. 46. 47. 48. 49. 50. 51. 52. 53. 54. 55.
 56. 57. 58. 59. 60. 61. 62. 63. 64. 65. 66. 67. 68. 69. 70. 71. 72. 73.
 74. 75. 76. 77. 78. 79. 80. 81. 82. 83. 84. 85. 86. 87. 88. 89. 90. 91.
 92. 93. 94.] | type: <class 'numpy.ndarray'> | ages.dtype: float64


In [10]:
# process.asr() | [1] Prepare spline.spline (knots, smoothing)
# ------------------------------------------------------------
knots = np.array(params_of_data_type.get('parameter_age_mesh', np.arange(ages[0], ages[-1] + 1, 5)), dtype=np.float64)

smooth_map = {'No Prior': np.inf, 'Slightly': 0.5, 'Moderately': 0.05, 'Very': 0.005}
smoothness_param = params_of_data_type.get('smoothness')

if isinstance(smoothness_param, dict): 
    amount = smoothness_param.get('amount')

    if isinstance(amount, (int, float)): # smoothness_param is dict, and amount is int or float
        smoothing = float(amount)
    else:                                # smoothness_param is dict, and amount may be string
        smoothing = smooth_map.get(amount, 0.0)

else:                                    # smoothness_param may be string
    smoothing = smooth_map.get(smoothness_param, 0.0)
# ------------------------------------------------------------
print(f'knots: {knots} | type: {type(knots)} | knots.dtype: {knots.dtype}')
print(f'smoothing: {smoothing}')

knots: [ 2. 30. 45. 60. 80. 94.] | type: <class 'numpy.ndarray'> | knots.dtype: float64
smoothing: 0.5


In [11]:
def inspect_model(model, var_name=None, show_shared_data=True):
    """
    Inspect a PyMC model. If var_name is None, print a summary,
    plus any shared_data contents. Otherwise, show details about a specific variable.
    """
    if var_name is None:
        print("📊 Model Summary:")
        print(f"  • Free RVs       : {len(model.free_RVs)} {[rv.name for rv in model.free_RVs]}")
        print(f"  • Observed RVs   : {len(model.observed_RVs)} {[rv.name for rv in model.observed_RVs]}")
        print(f"  • Deterministics : {len(model.deterministics)} {[rv.name for rv in model.deterministics]}")
        print(f"  • Potentials     : {len(model.potentials)} {[pot.name for pot in model.potentials]}")
        print(f"  • Total Named RVs: {len(model.named_vars)}")

        # --- Print shared_data contents if present ---
        if show_shared_data:
            if hasattr(model, "shared_data"):
                sd = model.shared_data
                if isinstance(sd, dict) and sd:
                    print("\n🔖 shared_data:")
                    for key, val in sd.items():
                        if isinstance(val, np.ndarray):
                            print(f"  • {key:15s}: array, shape={val.shape}, dtype={val.dtype}")
                        else:
                            print(f"  • {key:15s}: {val!r}")

    else:
        var_dict = model.named_vars
        if var_name not in var_dict:
            print(f"❌ Variable '{var_name}' not found in model.named_vars.")
            return

        var = var_dict[var_name]
        print(f"🔍 Variable: {var_name}")
        print(f"  • Type     : {type(var)}")
        print(f"  • Shape    : {getattr(var, 'shape', None)}")
        print(f"  • DType    : {getattr(var, 'dtype', None)}")
        print(f"  • Owner OP : {var.owner.op if getattr(var, 'owner', None) else 'None'}")

        if hasattr(var, 'distribution'):
            dist = var.distribution
            print(f"  • Distribution: {dist.__class__.__name__}")
            if hasattr(dist, 'dist'):
                print(f"    - PyMC Dist : {dist.dist.__class__.__name__}")
            if hasattr(dist, 'kwargs'):
                print("    - Parameters:")
                for k, v in dist.kwargs.items():
                    print(f"      {k}: {v}")

        if hasattr(var, 'eval'):
            try:
                val = var.eval()
                print(f"  • Current value (eval): {val}")
            except Exception as e:
                print(f"  • Could not evaluate variable: {e}")


In [12]:
pm_model = pm.Model()

with pm_model: 
    pm_model.shared_data = {    # NOTE: this is what used to be "vars" from class ModelVars
        "data_type": data_type,
        "ages":      ages,
        "age_weights": age_weights,
        "data":      data,
        "lb_data":   lb_data,
        "knots":     knots,
        "smoothing": smoothing,
        "interpolation_method": interpolation_method,
        "params_of_data_type": params_of_data_type,
        "reference_area_id": name_to_id[reference_area],
        "reference_sex": reference_sex,
        "reference_year": reference_year,
        "zero_re": zero_re,
        "region_id_graph": region_id_graph,
        "output_template": output_template,
    }

inspect_model(pm_model)

📊 Model Summary:
  • Free RVs       : 0 []
  • Observed RVs   : 0 []
  • Deterministics : 0 []
  • Potentials     : 0 []
  • Total Named RVs: 0

🔖 shared_data:
  • data_type      : 'p'
  • ages           : array, shape=(93,), dtype=float64
  • age_weights    : array, shape=(101,), dtype=float64
  • data           :                          area  location_id         stage  stage_id     sex   
0                 Netherlands           89  Intermediate         5    Male  \
1                 Netherlands           89  Intermediate         5  Female   
2                 Netherlands           89  Intermediate         5    Male   
3                 Netherlands           89  Intermediate         5  Female   
4                 Netherlands           89  Intermediate         5    Male   
..                        ...          ...           ...       ...     ...   
202                   Ireland           84  Intermediate         5    Both   
203                   Germany           81  Intermediate   

In [13]:
# [EXAMPLE] How to use shared_data in other functions
# ------------------------------------------------------------
with pm_model:

    # Get shared_data
    data_type = pm_model.shared_data["data_type"]
    ages      = pm_model.shared_data["ages"]
    data      = pm_model.shared_data["data"]
    lb_data   = pm_model.shared_data["lb_data"]
    knots     = pm_model.shared_data["knots"]
    smoothing = pm_model.shared_data["smoothing"]
# ------------------------------------------------------------

In [14]:
# process.asr() | (1) spline.py
import model.spline as spline
print(spline.__file__)

with pm_model:
    if mu_age is not None:
        unconstrained_mu_age_tv = mu_age
        
    else:
        unconstrained_mu_age_tv = spline.spline()
        
inspect_model(pm_model, show_shared_data=False)
print("-----------------------------------------------------")
inspect_model(pm_model, var_name='gamma_p_0', show_shared_data=False)

/Users/Dev/AMD/dismod_mr_migrated/reforged_mr/model/spline.py
printing type of gamma
<class 'list'>
[gamma_p_0, gamma_p_1, gamma_p_2, gamma_p_3, gamma_p_4, gamma_p_5]
printing type of gamma_vec
<class 'pytensor.tensor.var.TensorVariable'>
MakeVector{dtype='float64'}.0
printing type of exp_gamma
<class 'pytensor.tensor.var.TensorVariable'>
Elemwise{exp,no_inplace}.0
printing type of W_t
<class 'pytensor.tensor.var.TensorConstant'>
TensorConstant{[[1.      ..        ]]}
printing type of mu_age before pm.Deterministic
<class 'pytensor.tensor.var.TensorVariable'>
dot.0
printing type of mu_age after pm.Deterministic
<class 'pytensor.tensor.var.TensorVariable'>
dot.0
📊 Model Summary:
  • Free RVs       : 6 ['gamma_p_0', 'gamma_p_1', 'gamma_p_2', 'gamma_p_3', 'gamma_p_4', 'gamma_p_5']
  • Observed RVs   : 0 []
  • Deterministics : 1 ['mu_age_p']
  • Potentials     : 1 ['smooth_p']
  • Total Named RVs: 8
-----------------------------------------------------
🔍 Variable: gamma_p_0
  • Type     :

In [15]:
# process.asr() | (2) priors.py - level_constraints()
import model.priors as priors
print(priors.__file__)

with pm_model:
    constrained_mu_age_tv, unconstrained_mu_age_tv, parent_similarity_tv= \
        priors.level_constraints(unconstrained_mu_age_tv)
    

inspect_model(pm_model, show_shared_data=False)

/Users/Dev/AMD/dismod_mr_migrated/reforged_mr/model/priors.py
📊 Model Summary:
  • Free RVs       : 6 ['gamma_p_0', 'gamma_p_1', 'gamma_p_2', 'gamma_p_3', 'gamma_p_4', 'gamma_p_5']
  • Observed RVs   : 0 []
  • Deterministics : 2 ['mu_age_p', 'value_constrained_mu_age_p']
  • Potentials     : 2 ['smooth_p', 'parent_similarity_p']
  • Total Named RVs: 10


In [16]:
# process.asr() | (3) priors.py - derivative_constraints()
with pm_model:    
        mu_age_derivative_tv = priors.derivative_constraints(mu_age=constrained_mu_age_tv)
    
inspect_model(pm_model, show_shared_data=False)

📊 Model Summary:
  • Free RVs       : 6 ['gamma_p_0', 'gamma_p_1', 'gamma_p_2', 'gamma_p_3', 'gamma_p_4', 'gamma_p_5']
  • Observed RVs   : 0 []
  • Deterministics : 2 ['mu_age_p', 'value_constrained_mu_age_p']
  • Potentials     : 3 ['smooth_p', 'parent_similarity_p', 'mu_age_derivative_potential_p']
  • Total Named RVs: 11


In [17]:
# process.asr() | (4) priors.py - similar()
with pm_model:    
    if mu_age_parent is not None:
        parent_similarity_tv = priors.similar( # TODO: similar() is also used in level_constraints(). 
            mu_child=mu_age_derivative_tv,     #      Thus, it is hard to reduce parameters.
            mu_parent=mu_age_parent,           #      Moreover, concerns on pm.Potential(parent_similarity_tv)
            sigma_parent=sigma_age_parent,     #      What happens if it is called twice with same name?
            sigma_difference=0.0,
            offset=1e-9
        )

inspect_model(pm_model, show_shared_data=False)

📊 Model Summary:
  • Free RVs       : 6 ['gamma_p_0', 'gamma_p_1', 'gamma_p_2', 'gamma_p_3', 'gamma_p_4', 'gamma_p_5']
  • Observed RVs   : 0 []
  • Deterministics : 2 ['mu_age_p', 'value_constrained_mu_age_p']
  • Potentials     : 3 ['smooth_p', 'parent_similarity_p', 'mu_age_derivative_potential_p']
  • Total Named RVs: 11


In [18]:
# process.asr() | (5) age_groups.py - age_standardize_approx()
#               | (6) covariate.py - mean_covariate_model()
import model.age_groups as age_groups
import model.covariates as covariates
print(age_groups.__file__)
print(covariates.__file__)

with pm_model:    
    if len(data) > 0:
        data = data.copy()
        # 2-1) standard_error, effective_sample_size 채우기
        se = data['standard_error'].mask(
            data['standard_error'] < 0,
            (data['upper_ci'] - data['lower_ci']) / (2 * 1.96)
        )
        ess = data['effective_sample_size'].fillna(
            data['value'] * (1 - data['value']) / se**2
        )
        data['standard_error'] = se
        data['effective_sample_size'] = ess

        mu_interval_tv = age_groups.age_standardize_approx(mu_age=mu_age_derivative_tv)

        # 2-2) covariate & pi
        if include_covariates:
            pi_tv = covariates.mean_covariate_model(mu=mu_interval_tv)

        else:
            pi_tv = mu_interval_tv


inspect_model(pm_model, show_shared_data=False)

/Users/Dev/AMD/dismod_mr_migrated/reforged_mr/model/age_groups.py
/Users/Dev/AMD/dismod_mr_migrated/reforged_mr/model/covariates.py
📊 Model Summary:
  • Free RVs       : 46 ['gamma_p_0', 'gamma_p_1', 'gamma_p_2', 'gamma_p_3', 'gamma_p_4', 'gamma_p_5', 'sigma_alpha_p_0_z', 'sigma_alpha_p_1_z', 'sigma_alpha_p_2_z', 'sigma_alpha_p_3_z', 'sigma_alpha_p_4_z', 'alpha_p_31', 'alpha_p_56', 'alpha_p_62', 'alpha_p_64', 'alpha_p_70', 'alpha_p_71', 'alpha_p_65', 'alpha_p_67', 'alpha_p_68', 'alpha_p_69', 'alpha_p_100', 'alpha_p_102', 'alpha_p_73', 'alpha_p_81', 'alpha_p_83', 'alpha_p_84', 'alpha_p_86', 'alpha_p_89', 'alpha_p_92', 'alpha_p_137', 'alpha_p_138', 'alpha_p_142', 'alpha_p_158', 'alpha_p_159', 'alpha_p_163', 'alpha_p_164', 'alpha_p_4', 'alpha_p_5', 'alpha_p_6', 'alpha_p_8', 'alpha_p_9', 'alpha_p_18', 'beta_p_x_sdi', 'beta_p_x_tob', 'beta_p_x_sex']
  • Observed RVs   : 0 []
  • Deterministics : 10 ['mu_age_p', 'value_constrained_mu_age_p', 'cum_sum_mu_p', 'mu_interval_p', 'sigma_alpha_p_0'

In [19]:
# process.asr() | (7) covariate.py - mean_covariate_model()

with pm_model:    
    if len(data) <= 0:
        if include_covariates:
            pi_tv = covariates.mean_covariate_model(mu=None)


inspect_model(pm_model, show_shared_data=False)

📊 Model Summary:
  • Free RVs       : 46 ['gamma_p_0', 'gamma_p_1', 'gamma_p_2', 'gamma_p_3', 'gamma_p_4', 'gamma_p_5', 'sigma_alpha_p_0_z', 'sigma_alpha_p_1_z', 'sigma_alpha_p_2_z', 'sigma_alpha_p_3_z', 'sigma_alpha_p_4_z', 'alpha_p_31', 'alpha_p_56', 'alpha_p_62', 'alpha_p_64', 'alpha_p_70', 'alpha_p_71', 'alpha_p_65', 'alpha_p_67', 'alpha_p_68', 'alpha_p_69', 'alpha_p_100', 'alpha_p_102', 'alpha_p_73', 'alpha_p_81', 'alpha_p_83', 'alpha_p_84', 'alpha_p_86', 'alpha_p_89', 'alpha_p_92', 'alpha_p_137', 'alpha_p_138', 'alpha_p_142', 'alpha_p_158', 'alpha_p_159', 'alpha_p_163', 'alpha_p_164', 'alpha_p_4', 'alpha_p_5', 'alpha_p_6', 'alpha_p_8', 'alpha_p_9', 'alpha_p_18', 'beta_p_x_sdi', 'beta_p_x_tob', 'beta_p_x_sex']
  • Observed RVs   : 0 []
  • Deterministics : 10 ['mu_age_p', 'value_constrained_mu_age_p', 'cum_sum_mu_p', 'mu_interval_p', 'sigma_alpha_p_0', 'sigma_alpha_p_1', 'sigma_alpha_p_2', 'sigma_alpha_p_3', 'sigma_alpha_p_4', 'pi_p']
  • Potentials     : 8 ['smooth_p', 'parent_si

In [20]:
rate_types = ['beta_binom', 'binom', 'neg_binom', 'poisson', 'log_normal', 'normal', 'offset_log_normal']
rate_type = rate_types[6]

In [21]:
# process.asr() | (8) covariate.py - dispersion_covariate_model()
# process.asr() | (9) likelihood.py - neg_binom()
import model.likelihood as likelihood
print(likelihood.__file__)

with pm_model:
    if len(data) > 0:
        if rate_type == 'neg_binom':
            bad_ess = (data['effective_sample_size'] <= 0) | data['effective_sample_size'].isna()
            if bad_ess.any():
                data.loc[bad_ess, 'effective_sample_size'] = 0.0

            big_ess = data['effective_sample_size'] >= 1e10
            if big_ess.any():
                data.loc[big_ess, 'effective_sample_size'] = 1e10

            hetero = parameters.get('heterogeneity', None)
            lower = {'Slightly': 9.0, 'Moderately': 3.0, 'Very': 1.0}.get(hetero, 1.0)
            if data_type == 'pf':
                lower = 1e12

            delta_tv = covariates.dispersion_covariate_model(delta_lb=lower, delta_ub=lower * 9.0)

            likelihood.neg_binom(pi=pi_tv, delta=delta_tv)            


inspect_model(pm_model, show_shared_data=False)

/Users/Dev/AMD/dismod_mr_migrated/reforged_mr/model/likelihood.py
📊 Model Summary:
  • Free RVs       : 46 ['gamma_p_0', 'gamma_p_1', 'gamma_p_2', 'gamma_p_3', 'gamma_p_4', 'gamma_p_5', 'sigma_alpha_p_0_z', 'sigma_alpha_p_1_z', 'sigma_alpha_p_2_z', 'sigma_alpha_p_3_z', 'sigma_alpha_p_4_z', 'alpha_p_31', 'alpha_p_56', 'alpha_p_62', 'alpha_p_64', 'alpha_p_70', 'alpha_p_71', 'alpha_p_65', 'alpha_p_67', 'alpha_p_68', 'alpha_p_69', 'alpha_p_100', 'alpha_p_102', 'alpha_p_73', 'alpha_p_81', 'alpha_p_83', 'alpha_p_84', 'alpha_p_86', 'alpha_p_89', 'alpha_p_92', 'alpha_p_137', 'alpha_p_138', 'alpha_p_142', 'alpha_p_158', 'alpha_p_159', 'alpha_p_163', 'alpha_p_164', 'alpha_p_4', 'alpha_p_5', 'alpha_p_6', 'alpha_p_8', 'alpha_p_9', 'alpha_p_18', 'beta_p_x_sdi', 'beta_p_x_tob', 'beta_p_x_sex']
  • Observed RVs   : 0 []
  • Deterministics : 10 ['mu_age_p', 'value_constrained_mu_age_p', 'cum_sum_mu_p', 'mu_interval_p', 'sigma_alpha_p_0', 'sigma_alpha_p_1', 'sigma_alpha_p_2', 'sigma_alpha_p_3', 'sigma_

In [22]:
# process.asr() | (10) likelihood.py - log_normal()

with pm_model:
    if len(data) > 0:
        if rate_type == 'log_normal':
            missing = data['standard_error'] < 0
            if missing.any():
                data.loc[missing, 'standard_error'] = 1e6

            sigma_tv = pm.Uniform(
                name=f'sigma_{data_type}',
                lower=1e-4,
                upper=1.0,
            )

            likelihood.log_normal(pi=pi_tv, sigma=sigma_tv)


inspect_model(pm_model, show_shared_data=False)

📊 Model Summary:
  • Free RVs       : 46 ['gamma_p_0', 'gamma_p_1', 'gamma_p_2', 'gamma_p_3', 'gamma_p_4', 'gamma_p_5', 'sigma_alpha_p_0_z', 'sigma_alpha_p_1_z', 'sigma_alpha_p_2_z', 'sigma_alpha_p_3_z', 'sigma_alpha_p_4_z', 'alpha_p_31', 'alpha_p_56', 'alpha_p_62', 'alpha_p_64', 'alpha_p_70', 'alpha_p_71', 'alpha_p_65', 'alpha_p_67', 'alpha_p_68', 'alpha_p_69', 'alpha_p_100', 'alpha_p_102', 'alpha_p_73', 'alpha_p_81', 'alpha_p_83', 'alpha_p_84', 'alpha_p_86', 'alpha_p_89', 'alpha_p_92', 'alpha_p_137', 'alpha_p_138', 'alpha_p_142', 'alpha_p_158', 'alpha_p_159', 'alpha_p_163', 'alpha_p_164', 'alpha_p_4', 'alpha_p_5', 'alpha_p_6', 'alpha_p_8', 'alpha_p_9', 'alpha_p_18', 'beta_p_x_sdi', 'beta_p_x_tob', 'beta_p_x_sex']
  • Observed RVs   : 0 []
  • Deterministics : 10 ['mu_age_p', 'value_constrained_mu_age_p', 'cum_sum_mu_p', 'mu_interval_p', 'sigma_alpha_p_0', 'sigma_alpha_p_1', 'sigma_alpha_p_2', 'sigma_alpha_p_3', 'sigma_alpha_p_4', 'pi_p']
  • Potentials     : 8 ['smooth_p', 'parent_si

[DEBUG] data_type=p: p 배열에서 0 이하인 값 (총 3개):
    index=0, p[0]=[4.09026798e-02 3.38512764e-02 7.23589001e-02 3.69973190e-02
 1.03015075e-01 8.70431894e-02 2.04188482e-01 1.37423313e-01
 2.46435845e-01 2.40051348e-01 3.14121037e-01 3.22265625e-01
 2.66355140e-01 3.56666667e-01 3.88888889e-01 2.16494845e-01
 1.41388175e-02 1.23456790e-02 2.70270270e-02 2.33333333e-02
 4.30622010e-02 4.27251732e-02 1.04395604e-01 1.13065327e-01
 1.68918919e-02 1.83486239e-02 3.16455696e-02 2.49307479e-02
 6.45161290e-02 4.60122699e-02 1.28834356e-01 1.36094675e-01
 5.05050505e-03 3.96825397e-03 1.89573460e-02 2.76679842e-02
 1.30434783e-02 2.77777778e-02 2.38095238e-02 8.73786408e-02
 2.70270270e-02 1.42857143e-02 2.61780105e-02 2.18579235e-02
 3.63636364e-02 5.34759358e-02 1.14285714e-01 1.20481928e-01
 0.00000000e+00 9.80392157e-03 3.12500000e-02 9.70873786e-03
 5.00000000e-02 4.95049505e-02 1.48936170e-01 6.97674419e-02
 1.67803547e-01 1.61971831e-01 1.36134454e-01 1.31944444e-01
 1.66666667e-01 1.26353791e-01 6.95364238e-02 3.88888889e-02
 5.69105691e-02 3.79746835e-02 7.69230769e-02 7.81250000e-02
 1.18811881e-01 1.83908046e-01 2.33333333e-01 2.85714286e-01
 2.85714286e-01 0.00000000e+00 3.42465753e-04 7.11297071e-02
 3.11973019e-02 2.87816490e-02 1.56402737e-01 1.60818713e-01
 1.69481982e-01 1.28455285e-01 6.93641618e-02 8.09716599e-02
 3.79746835e-02 3.30578512e-02 6.97674419e-02 1.34831461e-01
 1.55555556e-01 7.90697674e-02 1.52380952e-01 4.54545455e-02
 5.69105691e-02 1.37254902e-01 1.47540984e-01 2.19178082e-01
...
 2.29729730e-01 6.79839577e-03 2.98547390e-02 7.06447188e-02
 1.10350982e-01 2.30137091e-02 2.93895112e-02 4.69371519e-02
 8.76531574e-02 1.32307692e-01 1.70212766e-01 8.61553785e-02
 1.98497854e-01 1.60000699e-02 1.60001603e-02]
Output is truncated. View as a scrollable element or open in a text editor. Adjust cell output settings...
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[47], line 16
      8                 data.loc[missing, 'standard_error'] = 1e6
     10             sigma_tv = pm.Uniform(
     11                 name=f'sigma_{data_type}',
     12                 lower=1e-4,
     13                 upper=1.0,
     14             )
---> 16             likelihood.log_normal(pi=pi_tv, sigma=sigma_tv)
     19 inspect_model(pm_model, show_shared_data=False)

File /Users/Dev/AMD/dismod_mr_migrated/reforged_mr/model/likelihood.py:378, in log_normal(pi, sigma)
    375         print(f"    index={i}, s[{i}]={s[i]}")
    377 # 2) 관측값 유효성 검사
--> 378 assert np.all(p > 0), 'observed values must be positive'
    379 assert np.all(s >= 0), 'standard error must be non-negative'
    381 # 3) 관측 로그값은 NumPy로 미리 계산

AssertionError: observed values must be positive

In [23]:
# process.asr() | (11) likelihood.py - log_normal()

with pm_model:
    if len(data) > 0:
        if rate_type == 'normal':
            missing = data['standard_error'] < 0
            if missing.any():
                data.loc[missing, 'standard_error'] = 1e6

            sigma_tv = pm.Uniform(
                name=f'sigma_{data_type}',
                lower=1e-4,
                upper=1e-1,
                initval=1e-2
            )

            likelihood.normal(pi=pi_tv, sigma=sigma_tv)


inspect_model(pm_model, show_shared_data=False)

📊 Model Summary:
  • Free RVs       : 46 ['gamma_p_0', 'gamma_p_1', 'gamma_p_2', 'gamma_p_3', 'gamma_p_4', 'gamma_p_5', 'sigma_alpha_p_0_z', 'sigma_alpha_p_1_z', 'sigma_alpha_p_2_z', 'sigma_alpha_p_3_z', 'sigma_alpha_p_4_z', 'alpha_p_31', 'alpha_p_56', 'alpha_p_62', 'alpha_p_64', 'alpha_p_70', 'alpha_p_71', 'alpha_p_65', 'alpha_p_67', 'alpha_p_68', 'alpha_p_69', 'alpha_p_100', 'alpha_p_102', 'alpha_p_73', 'alpha_p_81', 'alpha_p_83', 'alpha_p_84', 'alpha_p_86', 'alpha_p_89', 'alpha_p_92', 'alpha_p_137', 'alpha_p_138', 'alpha_p_142', 'alpha_p_158', 'alpha_p_159', 'alpha_p_163', 'alpha_p_164', 'alpha_p_4', 'alpha_p_5', 'alpha_p_6', 'alpha_p_8', 'alpha_p_9', 'alpha_p_18', 'beta_p_x_sdi', 'beta_p_x_tob', 'beta_p_x_sex']
  • Observed RVs   : 0 []
  • Deterministics : 10 ['mu_age_p', 'value_constrained_mu_age_p', 'cum_sum_mu_p', 'mu_interval_p', 'sigma_alpha_p_0', 'sigma_alpha_p_1', 'sigma_alpha_p_2', 'sigma_alpha_p_3', 'sigma_alpha_p_4', 'pi_p']
  • Potentials     : 8 ['smooth_p', 'parent_si

In [24]:
# process.asr() | (12) likelihood.py - binom()

with pm_model:
    if len(data) > 0:
        if rate_type == 'binom':
            bad_ess = data['effective_sample_size'] < 0
            if bad_ess.any():
                data.loc[bad_ess, 'effective_sample_size'] = 0.0

            likelihood.binom(pi=pi_tv)


inspect_model(pm_model, show_shared_data=False)

📊 Model Summary:
  • Free RVs       : 46 ['gamma_p_0', 'gamma_p_1', 'gamma_p_2', 'gamma_p_3', 'gamma_p_4', 'gamma_p_5', 'sigma_alpha_p_0_z', 'sigma_alpha_p_1_z', 'sigma_alpha_p_2_z', 'sigma_alpha_p_3_z', 'sigma_alpha_p_4_z', 'alpha_p_31', 'alpha_p_56', 'alpha_p_62', 'alpha_p_64', 'alpha_p_70', 'alpha_p_71', 'alpha_p_65', 'alpha_p_67', 'alpha_p_68', 'alpha_p_69', 'alpha_p_100', 'alpha_p_102', 'alpha_p_73', 'alpha_p_81', 'alpha_p_83', 'alpha_p_84', 'alpha_p_86', 'alpha_p_89', 'alpha_p_92', 'alpha_p_137', 'alpha_p_138', 'alpha_p_142', 'alpha_p_158', 'alpha_p_159', 'alpha_p_163', 'alpha_p_164', 'alpha_p_4', 'alpha_p_5', 'alpha_p_6', 'alpha_p_8', 'alpha_p_9', 'alpha_p_18', 'beta_p_x_sdi', 'beta_p_x_tob', 'beta_p_x_sex']
  • Observed RVs   : 0 []
  • Deterministics : 10 ['mu_age_p', 'value_constrained_mu_age_p', 'cum_sum_mu_p', 'mu_interval_p', 'sigma_alpha_p_0', 'sigma_alpha_p_1', 'sigma_alpha_p_2', 'sigma_alpha_p_3', 'sigma_alpha_p_4', 'pi_p']
  • Potentials     : 8 ['smooth_p', 'parent_si

In [25]:
# process.asr() | (13) likelihood.py - beta_binom()

with pm_model:
    if len(data) > 0:
        if rate_type == 'beta_binom':

            ### NEWLY ADDED: Origianl code doesn't have delta_tv calculation
            bad_ess = (data['effective_sample_size'] <= 0) | data['effective_sample_size'].isna()
            if bad_ess.any():
                data.loc[bad_ess, 'effective_sample_size'] = 0.0

            big_ess = data['effective_sample_size'] >= 1e10
            if big_ess.any():
                data.loc[big_ess, 'effective_sample_size'] = 1e10

            hetero = parameters.get('heterogeneity', None)
            lower = {'Slightly': 9.0, 'Moderately': 3.0, 'Very': 1.0}.get(hetero, 1.0)
            if data_type == 'pf':
                lower = 1e12

            delta_tv = covariates.dispersion_covariate_model(delta_lb=lower, delta_ub=lower * 9.0)

            likelihood.beta_binom(pi=pi_tv, delta=delta_tv)


inspect_model(pm_model, show_shared_data=False)

📊 Model Summary:
  • Free RVs       : 46 ['gamma_p_0', 'gamma_p_1', 'gamma_p_2', 'gamma_p_3', 'gamma_p_4', 'gamma_p_5', 'sigma_alpha_p_0_z', 'sigma_alpha_p_1_z', 'sigma_alpha_p_2_z', 'sigma_alpha_p_3_z', 'sigma_alpha_p_4_z', 'alpha_p_31', 'alpha_p_56', 'alpha_p_62', 'alpha_p_64', 'alpha_p_70', 'alpha_p_71', 'alpha_p_65', 'alpha_p_67', 'alpha_p_68', 'alpha_p_69', 'alpha_p_100', 'alpha_p_102', 'alpha_p_73', 'alpha_p_81', 'alpha_p_83', 'alpha_p_84', 'alpha_p_86', 'alpha_p_89', 'alpha_p_92', 'alpha_p_137', 'alpha_p_138', 'alpha_p_142', 'alpha_p_158', 'alpha_p_159', 'alpha_p_163', 'alpha_p_164', 'alpha_p_4', 'alpha_p_5', 'alpha_p_6', 'alpha_p_8', 'alpha_p_9', 'alpha_p_18', 'beta_p_x_sdi', 'beta_p_x_tob', 'beta_p_x_sex']
  • Observed RVs   : 0 []
  • Deterministics : 10 ['mu_age_p', 'value_constrained_mu_age_p', 'cum_sum_mu_p', 'mu_interval_p', 'sigma_alpha_p_0', 'sigma_alpha_p_1', 'sigma_alpha_p_2', 'sigma_alpha_p_3', 'sigma_alpha_p_4', 'pi_p']
  • Potentials     : 8 ['smooth_p', 'parent_si

---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Cell In[25], line 23
     19                 lower = 1e12
     21             delta_tv = covariates.dispersion_covariate_model(delta_lb=lower, delta_ub=lower * 9.0)
---> 23             likelihood.beta_binom(pi=pi_tv, delta=delta_tv)
     26 inspect_model(pm_model, show_shared_data=False)

File /Users/Dev/AMD/dismod_mr_migrated/reforged_mr/model/likelihood.py:275, in beta_binom(pi, delta)
    267 alpha_param = pi * delta * 50
    268 beta_param = (1 - pi) * delta * 50
    270 p_obs = pm.BetaBinomial(
    271     name=f'p_obs_{data_type}',
    272     n=n_int[mask],
    273     alpha=alpha_param[mask] if hasattr(alpha_param, 'shape') else alpha_param,
    274     beta=beta_param[mask] if hasattr(beta_param, 'shape') else beta_param,
--> 275     observed=obs_counts[mask]
    276 )
    278 # Posterior predictive counts: replace zero-sample cases
    279 n_pred = n_int.copy()

IndexError: boolean index did not match indexed array along dimension 0; dimension is 1 but corresponding boolean dimension is 207

In [26]:
# process.asr() | (14) likelihood.py - poisson()

with pm_model:
    if len(data) > 0:
        if rate_type == 'poisson':
            bad_ess = data['effective_sample_size'] < 0
            if bad_ess.any():
                data.loc[bad_ess, 'effective_sample_size'] = 0.0

            likelihood.poisson(pi=pi_tv)


inspect_model(pm_model, show_shared_data=False)

📊 Model Summary:
  • Free RVs       : 46 ['gamma_p_0', 'gamma_p_1', 'gamma_p_2', 'gamma_p_3', 'gamma_p_4', 'gamma_p_5', 'sigma_alpha_p_0_z', 'sigma_alpha_p_1_z', 'sigma_alpha_p_2_z', 'sigma_alpha_p_3_z', 'sigma_alpha_p_4_z', 'alpha_p_31', 'alpha_p_56', 'alpha_p_62', 'alpha_p_64', 'alpha_p_70', 'alpha_p_71', 'alpha_p_65', 'alpha_p_67', 'alpha_p_68', 'alpha_p_69', 'alpha_p_100', 'alpha_p_102', 'alpha_p_73', 'alpha_p_81', 'alpha_p_83', 'alpha_p_84', 'alpha_p_86', 'alpha_p_89', 'alpha_p_92', 'alpha_p_137', 'alpha_p_138', 'alpha_p_142', 'alpha_p_158', 'alpha_p_159', 'alpha_p_163', 'alpha_p_164', 'alpha_p_4', 'alpha_p_5', 'alpha_p_6', 'alpha_p_8', 'alpha_p_9', 'alpha_p_18', 'beta_p_x_sdi', 'beta_p_x_tob', 'beta_p_x_sex']
  • Observed RVs   : 0 []
  • Deterministics : 10 ['mu_age_p', 'value_constrained_mu_age_p', 'cum_sum_mu_p', 'mu_interval_p', 'sigma_alpha_p_0', 'sigma_alpha_p_1', 'sigma_alpha_p_2', 'sigma_alpha_p_3', 'sigma_alpha_p_4', 'pi_p']
  • Potentials     : 8 ['smooth_p', 'parent_si

In [27]:
# process.asr() | (15) likelihood.py - offset_log_normal()
#               | (16) else

with pm_model:
    if len(data) > 0:
        if rate_type == 'offset_log_normal':
            
            sigma_tv = pm.Uniform(
                name=f'sigma_{data_type}',
                lower=1e-4,
                upper=10.0,
                initval=1e-2
            )

            likelihood.offset_log_normal(pi=pi_tv, sigma=sigma_tv)

        else:
            raise ValueError(f'Unsupported rate_type "{rate_type}"')


inspect_model(pm_model, show_shared_data=False)

📊 Model Summary:
  • Free RVs       : 49 ['gamma_p_0', 'gamma_p_1', 'gamma_p_2', 'gamma_p_3', 'gamma_p_4', 'gamma_p_5', 'sigma_alpha_p_0_z', 'sigma_alpha_p_1_z', 'sigma_alpha_p_2_z', 'sigma_alpha_p_3_z', 'sigma_alpha_p_4_z', 'alpha_p_31', 'alpha_p_56', 'alpha_p_62', 'alpha_p_64', 'alpha_p_70', 'alpha_p_71', 'alpha_p_65', 'alpha_p_67', 'alpha_p_68', 'alpha_p_69', 'alpha_p_100', 'alpha_p_102', 'alpha_p_73', 'alpha_p_81', 'alpha_p_83', 'alpha_p_84', 'alpha_p_86', 'alpha_p_89', 'alpha_p_92', 'alpha_p_137', 'alpha_p_138', 'alpha_p_142', 'alpha_p_158', 'alpha_p_159', 'alpha_p_163', 'alpha_p_164', 'alpha_p_4', 'alpha_p_5', 'alpha_p_6', 'alpha_p_8', 'alpha_p_9', 'alpha_p_18', 'beta_p_x_sdi', 'beta_p_x_tob', 'beta_p_x_sex', 'sigma_p', 'p_zeta_p', 'log_pred_p']
  • Observed RVs   : 0 []
  • Deterministics : 11 ['mu_age_p', 'value_constrained_mu_age_p', 'cum_sum_mu_p', 'mu_interval_p', 'sigma_alpha_p_0', 'sigma_alpha_p_1', 'sigma_alpha_p_2', 'sigma_alpha_p_3', 'sigma_alpha_p_4', 'pi_p', 'p_pred_p

In [73]:
# Level 3: country
countries = (
    hier[hier["Level"] == 3]
    .rename(columns={
        "Location ID": "loc3_id",
        "Location Name": "country",
        "Parent ID": "parent2_id"
    })
    [["loc3_id", "country", "parent2_id"]]
)

# Level 2: region
regions = (
    hier[hier["Level"] == 2]
    .rename(columns={
        "Location ID": "loc2_id",
        "Location Name": "region",
        "Parent ID": "parent1_id"
    })
    [["loc2_id", "region", "parent1_id"]]
)

# Level 1: super-region
superregs = (
    hier[hier["Level"] == 1]
    .rename(columns={
        "Location ID": "loc1_id",
        "Location Name": "super_region"
    })
    [["loc1_id", "super_region"]]
)


NameError: name 'hier' is not defined

In [4]:
# 4.1 country 정보 붙이기
df = df.merge(countries, on="country", how="left")

# 4.2 region 정보 붙이기
df = df.merge(
    regions,
    left_on="parent2_id",
    right_on="loc2_id",
    how="left"
)

# 4.3 super-region 정보 붙이기
df = df.merge(
    superregs,
    left_on="parent1_id",
    right_on="loc1_id",
    how="left"
)

# 확인
df[["country","region","super_region"]].drop_duplicates().head()

country                     region super_region
0                    Sweden             Western Europe  High-income
1  United States of America  High-income North America  High-income
3                    Israel             Western Europe  High-income
5               Switzerland             Western Europe  High-income
7                    Canada  High-income North America  High-income

In [5]:

# 2) 매핑 딕셔너리
age_map  = {"Children":0, "Adult":1, "Both":2}
meth_map = {
    "Interview based":0,
    "Non-interview":1
}
prev_map = {
    "point prevalence":     0,
    "12-month prevalence":  1,
    "life-time prevalence": 2
}

# 3) long 포맷으로 전환
df_long = df.melt(
    id_vars=[
        "study",             
        "country",
        "region",
        "super_region",
        "Methods of questionnaire",
        "Age range",
        "study population"
    ],
    value_vars=[
        "point prevalence",
        "12-month prevalence",
        "life-time prevalence"
    ],
    var_name="prevalence_type",
    value_name="prevalence_value"
).dropna(subset=["prevalence_value"])

# 4) 코드 매핑
df_long["age_code"]  = df_long["Age range"].map(age_map)
df_long["meth_code"] = df_long["Methods of questionnaire"].map(meth_map)
df_long["prev_code"] = df_long["prevalence_type"].map(prev_map)

# 5) 최종 확인
result = df_long[[
    "study", "country", "region", "super_region",
    "age_code", "meth_code",
    "prev_code", "prevalence_value"
]]
print(result.head())

                     study                   country  \
0   Nilsson LV et al, 1984                    Sweden   
1     Kramer M et al, 1985  United States of America   
2    Flament M et al, 1990  United States of America   
3     Zohar AH et al, 1992                    Israel   
4  Reinherz HZ et al, 1993  United States of America   

                      region super_region  age_code  meth_code  prev_code  \
0             Western Europe  High-income         1          0          0   
1  High-income North America  High-income         1          0          0   
2  High-income North America  High-income         1          0          0   
3             Western Europe  High-income         0          0          0   
4  High-income North America  High-income         0          0          0   

   prevalence_value  
0              98.0  
1              22.0  
2              55.0  
3              36.0  
4              21.0  


In [6]:
# 6) 비율(p_hat) 및 로그 변환
#    prevalence_value: point/12-month/life-time 케이스 수
#    study population: 전체 표본 수

# 6.1 p_hat 계산
df_long["p_hat"] = df_long["prevalence_value"] / df_long["study population"]

# 6.2 로그 변환 (필요 시 0 회피를 위해 clamp)
eps = 1e-6
df_long["p_hat"] = df_long["p_hat"].clip(eps, 1 - eps)
df_long["log_p_hat"] = np.log(df_long["p_hat"])

In [ ]:
# 1) df_long 이용: prevalence_type 당 개수 세기
prev_count = df_long.groupby("study")["prevalence_type"] \
                    .transform("nunique")

# 2) 2개 이상이면 1, 아니면 0
df_long["multi_mask"] = (prev_count >= 2).astype(int)

# 3) 확인
df_long[["study","prevalence_type","multi_mask"]].drop_duplicates().head()

study   prevalence_type  multi_mask
0   Nilsson LV et al, 1984  point prevalence           0
1     Kramer M et al, 1985  point prevalence           0
2    Flament M et al, 1990  point prevalence           1
3     Zohar AH et al, 1992  point prevalence           0
4  Reinherz HZ et al, 1993  point prevalence           0

In [8]:
import pandas as pd

# 1) hierarchy 에서 좌표(리스트) 직접 추출
# level 1 → super_region (7개)
super_list = (
    hier[hier["Level"] == 1]
    ["Location Name"]
    .tolist()
)

# level 2 → region (21개)
region_list = (
    hier[hier["Level"] == 2]
    ["Location Name"]
    .tolist()
)

# level 3 → country (180+개)
country_list = (
    hier[hier["Level"] == 3]
    ["Location Name"]
    .tolist()
)

# 2) coords 정의 (unchanged)
coords = {
    "country":      country_list,
    "region":       region_list,
    "super_region": super_list,
    "age":          ["Children", "Adult", "Both"],
    "meth":         ["Interview based", "Non-interview"],
    "prev":         ["point prevalence", "12-month prevalence", "life-time prevalence"],
    "study":        df_long["study"].unique().tolist()
}

# 3) questionnaire method 컬럼명 변경 반영
#    이미 엑셀에서 "Methods of questionnaire" 아래 값을 "Interview based" / "Non-interview"로 바꿔 두셨다면
meth_map = {
    "Interview based":    0,
    "Non-interview":      1
}
df_long["meth_code"] = df_long["Methods of questionnaire"].map(meth_map)

# 4) index 배열 생성
country_idx = pd.Categorical(
    df_long["country"],
    categories=coords["country"]
).codes

region_idx = pd.Categorical(
    df_long["region"],
    categories=coords["region"]
).codes

super_idx = pd.Categorical(
    df_long["super_region"],
    categories=coords["super_region"]
).codes

age_idx   = df_long["age_code"].to_numpy()
meth_idx  = df_long["meth_code"].to_numpy()         
prev_idx  = df_long["prev_code"].to_numpy()

study_idx = pd.Categorical(
    df_long["study"],
    categories=coords["study"]
).codes

# 5) mask / 관측치 벡터 (unchanged)
multi_mask = df_long["multi_mask"].to_numpy()
y_obs      = df_long["log_p_hat"].to_numpy()


In [ ]:


with pm.Model(coords=coords) as model:

    # ─── Fixed effects ───
    β0    = pm.Normal("β0",    mu=0, sigma=10)
    β_age = pm.Normal("β_age", mu=0, sigma=5, dims="age")
    β_meth= pm.Normal("β_meth",mu=0, sigma=5, dims="meth")
    β_prev= pm.Normal("β_prev",mu=0, sigma=5, dims="prev")

    # ─── Non-centered random intercepts ───
    σ_sr = pm.HalfNormal("σ_sr", sigma=5)
    η_sr = pm.Normal("η_sr", mu=0, sigma=1, dims="super_region")
    u_sr = pm.Deterministic("u_sr", η_sr * σ_sr)

    σ_r  = pm.HalfNormal("σ_r", sigma=5)
    η_r  = pm.Normal("η_r",  mu=0, sigma=1, dims="region")
    u_r  = pm.Deterministic("u_r", η_r * σ_r)

    σ_c  = pm.HalfNormal("σ_c", sigma=5)
    η_c  = pm.Normal("η_c",  mu=0, sigma=1, dims="country")
    u_c  = pm.Deterministic("u_c", η_c * σ_c)

    # ─── Study-level random slope (non-centered) ───
    σ_ps= pm.HalfNormal("σ_prev_study", sigma=5)
    η_ps= pm.Normal("η_prev_study", mu=0, sigma=1, dims=("study","prev"))
    u_ps= pm.Deterministic("u_prev_study", η_ps * σ_ps)

    # ─── 선형 예측식 μ ───
    prev_effect = (β_prev[prev_idx] + u_ps[study_idx,prev_idx]) * multi_mask

    μ = (
        β0
        + β_age[age_idx]
        + β_meth[meth_idx]
        + prev_effect
        + u_sr[super_idx]
        + u_r[region_idx]
        + u_c[country_idx]
    )

    # ─── 관측모형 ───
    σ_y = pm.HalfNormal("σ_y", sigma=1)
    pm.Normal("y", mu=μ, sigma=σ_y, observed=y_obs)

    # ─── (옵션) ADVI 워밍업 — 빠르게 근사 posterior 생성 ───
    advi = pm.fit(method="advi", n=5000)

    # ─── NUTS 샘플링 ───
    idata = pm.sample(
        draws=1000, tune=500,
        chains=4, cores=4,
        target_accept=0.95,
        max_treedepth=15,
        progressbar="split"
    )

# ─── forest plot 으로 깔끔하게 시각화 ───
az.style.use("arviz-darkgrid")
az.plot_forest(
    idata,
    var_names=["β_age","β_meth","β_prev","σ_sr","σ_r","σ_c","σ_prev_study","σ_y"],
    combined=True,
    credible_interval=0.95,
    ridgeplot_overlap=0.5
)


Output()

Finished [100%]: Average Loss = 395.52
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [β0, β_age, β_meth, β_prev, σ_sr, η_sr, σ_r, η_r, σ_c, η_c, σ_prev_study, η_prev_study, σ_y]


Output()

Sampling 4 chains for 500 tune and 1_000 draw iterations (2_000 + 4_000 draws total) took 7188 seconds.


TypeError: plot_forest() got an unexpected keyword argument 'credible_interval'

In [12]:
import arviz as az

# idata는 이미 pm.sample()의 결과
summary_df = az.summary(
    idata,
    var_names=[
        "β0", "β_age", "β_meth", "β_prev",
        "σ_sr", "σ_r", "σ_c", "σ_prev_study", "σ_y"
    ],
    round_to=3,         # 소수점 자리수
    hdi_prob=0.95       # 95% 신뢰구간
)
print(summary_df)


                               mean     sd  hdi_2.5%  hdi_97.5%  mcse_mean  \
β0                           -3.537  4.085   -11.136      4.469      0.111   
β_age[Children]              -0.263  2.759    -5.733      4.756      0.080   
β_age[Adult]                 -0.284  2.752    -5.538      4.946      0.080   
β_age[Both]                  -0.183  2.763    -5.575      4.984      0.080   
β_meth[Interview based]       0.241  3.283    -6.431      6.356      0.080   
β_meth[Non-interview]        -0.924  3.280    -7.457      5.359      0.080   
β_prev[point prevalence]     -1.014  0.388    -1.760     -0.244      0.006   
β_prev[12-month prevalence]  -0.401  0.291    -0.957      0.177      0.006   
β_prev[life-time prevalence] -0.121  0.290    -0.680      0.434      0.006   
σ_sr                          0.916  0.575     0.008      1.955      0.017   
σ_r                           0.291  0.242     0.000      0.755      0.008   
σ_c                           0.455  0.218     0.011      0.816 

In [13]:


# 1) posterior를 (chain, draw) → sample 차원 하나로 합치기
post = idata.posterior.stack(sample=("chain","draw"))

# 2) coords에 저장된 리스트 불러오기
country_list      = coords["country"]       # Level-3 국가 이름 리스트
super_list        = coords["super_region"]  # Level-1 슈퍼리전
region_list       = coords["region"]        # Level-2 리전
age_list          = coords["age"]           # ["Children","Adult","Both"]
meth_list         = coords["meth"]          # ["Interview based","Non-interview"]
prev_list         = coords["prev"]          # ["point prevalence","12-month prevalence","life-time prevalence"]

# 3) 관심 범주의 posterior 샘플 꺼내기
β0_samps         = post["β0"].values                             # (samples,)
β_age_Both       = post["β_age"].sel(age="Both").values          # (samples,)
β_meth_IB        = post["β_meth"].sel(meth="Interview based").values  # (samples,)
β_prev_life      = post["β_prev"].sel(prev="life-time prevalence").values  # (samples,)

u_sr_samps       = post["u_sr"].values      # (super_region, samples)
u_r_samps        = post["u_r"].values       # (region, samples)
u_c_samps        = post["u_c"].values       # (country, samples)

# 4) country → region, super_region 매핑 (levels 파일에서 미리 만들어 두었다면 그것 사용)
#    여기선 hier 데이터프레임에 있는 Level, Location Name, Parent ID 이용
mapping = hier[["Level","Location ID","Location Name","Parent ID"]]
# Level3
lvl3 = mapping[mapping["Level"]==3].rename(columns={
    "Location Name":"country","Location ID":"loc3","Parent ID":"parent2"
})
# Level2
lvl2 = mapping[mapping["Level"]==2].rename(columns={
    "Location Name":"region","Location ID":"loc2","Parent ID":"parent1"
})
# Level1
lvl1 = mapping[mapping["Level"]==1].rename(columns={
    "Location Name":"super_region","Location ID":"loc1"
})

# merge to get each country’s region & super_region names
df_map = (
    lvl3[["country","loc3","parent2"]]
    .merge(lvl2[["loc2","region","parent1"]], left_on="parent2", right_on="loc2", how="left")
    .merge(lvl1[["loc1","super_region"]], left_on="parent1", right_on="loc1", how="left")
    [["country","region","super_region"]]
)

# 5) Monte Carlo 샘플마다 μ 계산 → p = exp(μ)
results = []
n_samps = β0_samps.shape[0]
for idx, row in df_map.iterrows():
    country    = row["country"]
    region     = row["region"]
    super_reg  = row["super_region"]
    i_sr = super_list.index(super_reg)
    i_r  = region_list.index(region)
    i_c  = country_list.index(country)

    # μ 샘플 벡터
    mu = (
        β0_samps
        + β_age_Both
        + β_meth_IB
        + β_prev_life
        + u_sr_samps[i_sr, :]
        + u_r_samps[i_r, :]
        + u_c_samps[i_c, :]
    )
    # prevalence 비율로 변환
    p = np.exp(mu)

    # median & 95% HDI
    med = np.median(p)
    hdi_lo, hdi_hi = az.hdi(p, hdi_prob=0.95)

    results.append({
        "country":             country,
        "region":              region,
        "super_region":        super_reg,
        "median_life_prev":    med,
        "hdi_2.5%":            hdi_lo,
        "hdi_97.5%":           hdi_hi
    })

df_out = pd.DataFrame(results)

# 6) 엑셀로 저장
output_path = "C:/Users/정이든/Desktop/연구실/9. prevalence of OCD/3차 와꾸/life_time_prevalence.xlsx"
df_out.to_excel(output_path, index=False)
print(f"Saved to {output_path}")


Saved to C:/Users/정이든/Desktop/연구실/9. prevalence of OCD/3차 와꾸/life_time_prevalence.xlsx
